include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+                                   data.match_step ?
+                                     TS_EXACTFINALTIME_MATCHSTEP :
+                                     TS_EXACTFINALTIME_STEPOVER));
+
+      // Adaptive tolerances
+      PetscReal atol = data.absolute_tolerance > 0.0 ?
+                         data.absolute_tolerance :
+                         static_cast<PetscReal>(PETSC_DEFAULT);
+      PetscReal rtol = data.relative_tolerance > 0.0 ?
+                         data.relative_tolerance :
+                         static_cast<PetscReal>(PETSC_DEFAULT);
+      AssertTS(TSSetTolerances(ts, atol, nullptr, rtol, nullptr));
+
+      // Adaptive time stepping
+      TSAdapt tsadapt;
+      AssertTS(TSGetAdapt(ts, &tsadapt));
+
+      // Use adaptive time stepping only if explicitly requested
+      if (data.absolute_tolerance == 0.0 && data.relative_tolerance == 0.0)
+        {
+          AssertTS(TSAdaptSetType(tsadapt, TSADAPTNONE));
+        }
+      else
+        {
+          AssertTS(TSAdaptSetType(tsadapt, TSADAPTBASIC));
+        }
+
+      // Step limits
+      PetscReal hmin = data.minimum_step_size > 0 ?
+                         data.minimum_step_size :
+                         static_cast<PetscInt>(PETSC_DEFAULT);
+      PetscReal hmax = data.maximum_step_size > 0 ?
+                         data.maximum_step_size :
+                         static_cast<PetscInt>(PETSC_DEFAULT);
+      AssertTS(TSAdaptSetStepLimits(tsadapt, hmin, hmax));
+    }
+
+
+    /**
+     * Pass the preconditioning matrix only.
+     * When used with 'setup_jacobian' and 'solve_for_jacobian_system', PETSc
+     * will approximate the linear system matrix-vector product using an
+     * internal matrix-free representation.
+     * When used with 'implicit_jacobian' or 'explicit_jacobian',
+     * PETSc will use the same matrix for both preconditioning and
+     * matrix-vector products.
+     */
+    void
+    reinit_matrices(PMatrixType &P)
+    {
+      this->A = nullptr;
+      this->P = &P;
+    }
+
+    /**
+     * Pass both the linear system matrix and the preconditioning matrix.
+     */
+    void
+    reinit_matrices(AMatrixType &A, PMatrixType &P)
+    {
+      this->A = &A;
+      this->P = &P;
+    }
+
+    /**
+     * Return current time
+     */
+    double
+    get_time()
+    {
+      PetscReal t;
+      AssertTS(TSGetTime(ts, &t));
+      return t;
+    }
+
+    /**
+     * Return current time step
+     */
+    double
+    get_time_step()
+    {
+      PetscReal dt;
+      AssertTS(TSGetTimeStep(ts, &dt));
+      return dt;
+    }
+
+    /**
+     * Integrate the differential-algebraic equations.
+     * This function returns the final number of computed steps.
+     */
+    unsigned int
+    solve(VectorType &solution)
+    {
+      auto _dealii_ts_ifunction_ =
+        [](TS ts, PetscReal t, Vec x, Vec xdot, Vec f, void *ctx)
+        -> PetscErrorCode {
+        TimeStepper *myctx;
+
+        PetscFunctionBeginUser;
+        (void)ctx;
```

Comment:
If you don't need the `ctx` argument, you can just do the following:
```suggestion
        [](TS ts, PetscReal t, Vec x, Vec xdot, Vec f, void *)
        -> PetscErrorCode {
        TimeStepper *myctx;

        PetscFunctionBeginUser;
```

Out of curiosity, why does TS pass a context object if one then needs to ignore that argument and instead obtain the context from the `ts` object?

Reply:
I see. I have no preference for which context object you use -- use what's idiomatic PETSc style in this regard.